core/src/lib.rs

```diff
@@ -1,64 +1,145 @@
-use parser::Element;
+use std::str::FromStr;
 
 mod context;
+mod element;
 mod error;
 mod package;
 
 pub use context::Context;
+pub use element::Element;
 pub use error::CoreError;
-pub use package::{Arg, NodeName, Package, PackageInfo, Transform};
+pub use package::{ArgInfo, Package, PackageInfo, Transform};
+use serde::Deserialize;
+use std::hash::{Hash, Hasher};
 
 #[cfg(all(feature = "web", feature = "native"))]
 compile_error!("feature \"native\" and feature \"web\" cannot be enabled at the same time");
 
+#[derive(Debug, Clone, Eq, Deserialize)]
+pub struct OutputFormat(String);
+
+impl OutputFormat {
+    pub fn new(string: &str) -> Self {
+        OutputFormat(string.to_lowercase())
+    }
+}
+
+/// To ensure that "html" and "HTML" is the same.
+impl PartialEq for OutputFormat {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.to_lowercase() == other.0.to_lowercase()
+    }
+}
+
+impl Hash for OutputFormat {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.to_lowercase().hash(state);
+    }
+}
+
+impl ToString for OutputFormat {
+    fn to_string(&self) -> String {
+        self.0.to_lowercase()
+    }
+}
+
+impl FromStr for OutputFormat {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(OutputFormat::new(s))
+    }
+}
+
 /// Evaluates a document using the given context
-pub fn eval(_document: &Element, _ctx: &mut Context) -> String {
-    "TODO".to_string()
+pub fn eval(source: &str, ctx: &mut Context, format: &OutputFormat) -> Result<String, CoreError> {
+    let document = parser::parse(source)?.try_into()?;
+    eval_elem(document, ctx, format)
+}
+
+pub fn eval_elem(
+    root: Element,
+    ctx: &mut Context,
+    format: &OutputFormat,
+) -> Result<String, CoreError> {
+    use Element::*;
+    match root {
+        Parent {
+            name: _,
+            args: _,
+            children: _,
+        } => {
+            let compound = ctx.transform(&root, format)?;
+            eval_elem(compound, ctx, format)
+        }
+        Compound(children) => {
+            let mut raw_content = String::new();
+
+            for child in children {
+                raw_content.push_str(&eval_elem(child, ctx, format)?);
+            }
+            Ok(raw_content)
+        }
+        Module {
+            ref name,
+            args: _,
+            ref body,
+            inline: _,
+        } => {
+            // Base case, if its just raw content, stop.
+            if name == "raw" {
+                return Ok(body.clone());
+            }
+
+            if name == "inline_content" {
```

Comment:
This isn't something that needs do be done now, but I think we would want to move the "native" modules out to a separate file since we will need to add some more later on, and add some proper manifests to them. It would be nice if we could use the same `Module` struct as before, and have an enum for the implementation, if it is implemented in wasm or natively by us, to make it more uniform

Reply:
Good suggestion, I've created GH-82.